### PR TITLE
Add cheat picker window

### PR DESCRIPTION
**Pull request by OFFTKP**

### CODE REVIEW

**Comment by OFFTKP** [comment]
image

**Comment by wheremyfoodat** [comment]
Beautiful menu
However there's a few problems:
- The code is not thread-safe at all. It should properly use the message queue like eg the Lua script or load rom actions.
-  It calls setEnable on the cheats window from a non-UI thread which I'm not sure is legal?
- Default build type should probably be release since it's tuned for non-Panda-devs that just want to build their own copy or slightly modified copy without thinking too much about it.

File: src/panda_qt/cheats_window.cpp [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+#include <QDialog>
+#include <QLineEdit>
+#include <QTextEdit>
+#include <QDialogButtonBox>
+#include "cheats.hpp"
+#include "emulator.hpp"
+
+using CheatHandle = u32;
+
+CheatHandle BAD_CHEAT = 0xFFFFFFFF;
```
Comment (wheremyfoodat) [inline src/panda_qt/cheats_window.cpp:17]: constexpr

File: src/panda_qt/cheats_window.cpp [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+
+	Cheats::Cheat cheat;
+	cheat.enabled = true;
+	cheat.type = Cheats::CheatType::ActionReplay;
+
+	for (size_t i = 0; i < size; i += 8) {
+		auto read32 = [](const u8* ptr) { return (u32(ptr[3]) << 24) | (u32(ptr[2]) << 16) | (u32(ptr[1]) << 8) | u32(ptr[0]); };
+
+		// Data is passed to us in big endian so we bswap
+		u32 firstWord = Common::swap32(read32(data + i));
```
Comment (wheremyfoodat) [inline src/panda_qt/cheats_window.cpp:41]: I am surprised this compiles without including swap.hpp

Comment (wheremyfoodat) [reply src/panda_qt/cheats_window.cpp:41]: (IWYU bby)

**Comment by OFFTKP** [comment]
> Beautiful menu However there's a few problems:
> 
>     * The code is not thread-safe at all. It should properly use the message queue like eg the Lua script or load rom actions.
> 
>     * It calls setEnable on the cheats window from a non-UI thread which I'm not sure is legal?
> 
>     * Default build type should probably be release since it's tuned for non-Panda-devs that just want to build their own copy or slightly modified copy without thinking too much about it.

* Now using message queue
* Removed setEnabled for now, will probably need to use QMetaObject::invokeMethod to run code on the GUI thread
* Yup my bad

**Comment by wheremyfoodat** [comment]
Boop

File: include/panda_qt/main_window.hpp [hunk 3]
```diff
@@ -12,16 +12,24 @@
...
+#include "panda_qt/cheats_window.hpp"
 #include "panda_qt/screen.hpp"
 #include "panda_qt/text_editor.hpp"
 #include "services/hid.hpp"
 
+struct CheatMessage
+{
+	u32 handle;
+	std::vector<uint8_t> cheat;
+	std::function<void(u32)> callback;
```
Comment (wheremyfoodat) [inline include/panda_qt/main_window.hpp:24]: (<functional> IWYU)

**Comment by XD89k** [comment]
Hola bro como estas soy un programador principiante quiero dedicarme a los emuladores no sé si estarías dispuesto a enseñarme un poco

**Comment by OFFTKP** [comment]
Hola. Le sugiero que visite el subreddit r/Emudev y se una al servidor Discord de emudev:
https://discord.gg/XD3bCwts

es un servidor en ingles

**Pull request merged**